presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/SpatialPartitioningStateFactory.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.plugin.geospatial;
+
+import com.esri.core.geometry.Envelope;
+import com.facebook.presto.array.IntBigArray;
+import com.facebook.presto.array.ObjectBigArray;
+import com.facebook.presto.geospatial.Rectangle;
+import com.facebook.presto.spi.function.AccumulatorStateFactory;
+import com.facebook.presto.spi.function.GroupedAccumulatorState;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.lang.Math.toIntExact;
+
+public class SpatialPartitioningStateFactory
+        implements AccumulatorStateFactory<SpatialPartitioningState>
+{
+    @Override
+    public SpatialPartitioningState createSingleState()
+    {
+        return new SingleSpatialPartitioningState();
+    }
+
+    @Override
+    public Class getSingleStateClass()
+    {
+        return SpatialPartitioningState.class;
+    }
+
+    @Override
+    public SpatialPartitioningState createGroupedState()
+    {
+        return new GroupedSpatialPartitioningState();
+    }
+
+    @Override
+    public Class getGroupedStateClass()
+    {
+        return GroupedSpatialPartitioningState.class;
+    }
+
+    public static final class GroupedSpatialPartitioningState
+            implements GroupedAccumulatorState, SpatialPartitioningState
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedSpatialPartitioningState.class).instanceSize();
+        private static final int ENVELOPE_SIZE = toIntExact(new Envelope(1, 2, 3, 4).estimateMemorySize());
+
+        private long groupId;
+        private IntBigArray partitionCounts = new IntBigArray();
+        private IntBigArray counts = new IntBigArray();
+        private ObjectBigArray<Rectangle> envelopes = new ObjectBigArray<>();
```

Comment:
Create an internal class, called, say `SingleState`. Make `SingleState` to include `partitionCount`, `count`, `extent`, `samples`. Than use a single `ObjectBigArray<SingleState>` here.